Пул-реквест для сдачи ДЗ

koroliana opened the pull request:
Добавлено меню волонтера и новые команды в enum	
Переработано хранение состояний (Очередь в json)
Создан класс для константных строк
Миграция БД с созданием таблиц кошек и собак
Покрыты тестами хэндлеры из меню консультации по подготовке ко встрече с новым питомцем (/adopt)

Microd18 committed:
Добавил меню волонтера и обновил список команд

Pruglo92 committed:
#26 Переработать подход для хранения состояний, состояния хранятся в виде json

Pruglo92 committed:
Merge branch 'develop' into #26

Pruglo92 committed:
Merge pull request #37 from Microd18/#26

#26

DmitriiVolkovIzh committed:
Тесты handlers - 1 и создал класс для константных строк MessagesBot

DmitriiVolkovIzh committed:
Тесты handlers - 1 и создал класс для константных строк MessagesBot

DmitriiVolkovIzh committed:
Merge remote-tracking branch 'origin/develop' into #29

Pruglo92 committed:
#32 Создать миграцию бд с созданием таблиц кошек и собак

Pruglo92 committed:
#32 Создать миграцию бд с созданием таблиц кошек и собак FIX

Pruglo92 committed:
#32 Создать миграцию бд с созданием таблиц кошек и собак FIX 2.0

DmitriiVolkovIzh committed:
Тесты handlers - 1 и создал класс для константных строк MessagesBot

Pruglo92 committed:
Merge pull request #39 from Microd18/#32

#32

DmitriiVolkovIzh committed:
Тест для доков

DmitriiVolkovIzh committed:
Часть тестов и константных текстовых команд

DmitriiVolkovIzh committed:
Merge pull request #38 from Microd18/#29

Часть тестов, константные текстовые команды в отдельном классе

koroliana committed:
изменен telegram.bot.token

Microd18 committed:
Исправил последнюю команду в меню /admin

koroliana committed:
реализованы 2 команды в меню волонтера: /find_user_by_phone и /find_animal_by_name

Pruglo92 committed:
#30 Сохранить отчет от пользователя о животном в БД

Pruglo92 committed:
Merge branch 'develop' into #30

# Conflicts:
#	src/test/java/pro/sky/AnimalShelter/service/ChatStateServiceTest.java

Pruglo92 committed:
Merge pull request #41 from Microd18/#30

#30

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/entity/ChatState.java:
```diff
@@ -16,34 +18,21 @@
-    private BotCommand currentState;
+    @Type(type = "json")
+    @Column(name = "state_data", columnDefinition = "jsonb")
+    private String stateData;
```
здесь можно писать прям класс свой, который описывает JSON, а не строку

Pruglo92 replied:
Про подход с классом узнал позже, когда уже реализовали через мапу

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/entity/UserReportState.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Type(type = "json")
+    @Column(name = "state_data", columnDefinition = "jsonb")
+    private String stateData;
```
лучше использовать класс свой вместо строки

Pruglo92 replied:
Про подход с классом узнал позже, когда уже реализовали через мапу

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/service/ChatStateService.java:
```diff
@@ -52,9 +71,21 @@ public BotCommand getCurrentStateByChatId(Long chatId) {
+                .map(stateQueue -> {
+                    if (!stateQueue.isEmpty()) {
+                        Iterator<BotCommand> iterator = stateQueue.descendingIterator();
+                        iterator.next();
```
почему вызов next до проверки происходит?

Pruglo92 replied:
не совсем понял вопрос, сперва проверяется, что очередь не пустая if (!stateQueue.isEmpty()), затем создаётся обратный итератор Iterator<BotCommand> iterator = stateQueue.descendingIterator();, только потом у него вызывается метод next iterator.next(); , то-есть если коллекция не пустая то в ней есть хотя-бы один элемент и метод iterator.next(); отработает корректно, разве нет?

dmitry-bizin replied:
я имел ввиду hasNext, но проглядел выше проверку на пустоту списка, так что всё норм)

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/service/UserReportService.java:
```diff
@@ -0,0 +1,274 @@
+                if (report instanceof DogReport) {
+                    DogReport dogReport = (DogReport) report;
+                    dogReport.setUser(user);
+                    dogReport.setDog(foundDog.get());
```
можно было бы сделать через полиморфизм - назвать метод setAnimal и сетить туда собаку или кошку, только надо завести базовый класс для собаки и кошки

dmitry-bizin replied:
тогда кода было бы в 2 раза меньше и не было бы каста и ужасного instanceof)

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/service/UserReportService.java:
```diff
@@ -0,0 +1,274 @@
+     * @return Массив байтов данных изображения.
+     */
+    private byte[] photoSizeToBytes(PhotoSize photoSize) {
+        log.info("photoSizeToBytes method was invoked");
+        String fileUrl = "https://api.telegram.org/file/bot" + telegramBot.getToken() + "/"
+                + telegramBot.execute(new GetFile(photoSize.fileId())).file().filePath();
+        try (InputStream inputStream = new java.net.URL(fileUrl).openStream()) {
+            return inputStream.readAllBytes();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
```
скачивать лучше через библиотеку как я показывал на разборе, а не руками

Pruglo92 replied:
Согласен, поправил

dmitry-bizin commented on src/main/java/pro/sky/AnimalShelter/utils/EnumDequeDeserializer.java:
```diff
@@ -0,0 +1,64 @@
+     * @throws IOException Возникает, если возникают ошибки при чтении JSON или десериализации.
+     */
+    @Override
+    public Deque<T> deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
```
а зачем так сложно? почему нельзя воспользоваться ObjectMapper-ом. И в конечном счёте сериализацию можно настроить, чтобы она происходила [автоматом](https://vladmihalcea.com/how-to-map-json-objects-using-generic-hibernate-types/)

Pruglo92 replied:
От этого подхода уже отказались, просто задублировали методы для обоих типов енама, так как времени на обдумывание и поиск инфы уже нету давным давно

dmitry-bizin commented on src/test/java/pro/sky/AnimalShelter/shelterInfoMenuHandlers/ScheduleCommandHandlerTest.java:
```diff
@@ -0,0 +1,97 @@
+        when(chatStateService.getCurrentStateByChatId(chatId)).thenReturn(currentState);
+        SendMessage message = new SendMessage(chatId.toString(), SCHEDULE_COMMAND_TEXT);
+        telegramBot.execute(message);
+        verify(telegramBot, times(1)).execute(any(SendMessage.class));
```
хорошо бы проверять во всех тестах, а какое именно отправляется сообщение, его содержимое, потому что сейчас у вас очень слабые проверки

Pruglo92 replied:
это исправим без проблем

Pruglo92 committed:
#29 HotFix тестов обработчиков

Pruglo92 committed:
Merge pull request #42 from Microd18/#29HotFix

#29 HotFix тестов обработчиков

koroliana committed:
реализована команда /make_adopter по записи питомцу user_id

koroliana committed:
telegram.bot.token возвращен в исходное состояние

koroliana committed:
Пофиксила выход за пределы массива при делении строчки на подстрочки

koroliana committed:
Merge branch 'develop' into koroliana#28

koroliana committed:
Merge pull request #43 from #28 Реализовать перевод юзера в усыновителя 

#28 Реализовать перевод юзера в усыновителя

koroliana committed:
Вернула недостающие импорты

koroliana committed:
Merge pull request #44 from Microd18/koroliana#hotfix

Вернула недостающие импорты

Pruglo92 committed:
#34 Покрыть тестами все сервисы которые ещё не покрыты
(покрыт UserReportStateService)

Pruglo92 committed:
Merge pull request #45 from Microd18/#34

#34 Покрыть тестами все сервисы которые ещё не покрыты

DmitriiVolkovIzh committed:
Тест adopt и админа

DmitriiVolkovIzh committed:
Тест adopt и админа 1.1.

DmitriiVolkovIzh committed:
Merge pull request #46 from Microd18/#29.1

#29.1

DmitriiVolkovIzh committed:
Хэнд админа 1.2.

DmitriiVolkovIzh committed:
Хэнд админа 1.2.1

DmitriiVolkovIzh committed:
Merge pull request #47 from Microd18/#29.1

#29.1

DmitriiVolkovIzh committed:
Покрытие тестами по меню adopt на 100%

DmitriiVolkovIzh committed:
Покрытие тестами по меню adopt на 100%

DmitriiVolkovIzh committed:
Покрытие тестов по строкам 54%

DmitriiVolkovIzh committed:
Merge pull request #48 from Microd18/#29

#29

DmitriiVolkovIzh committed:
дополнение тестов шелтеринфо и меню волонтера

DmitriiVolkovIzh committed:
Merge pull request #50 from Microd18/#29

дополнение тестов шелтеринфо и меню волонтера

koroliana committed:
Merge remote-tracking branch 'origin/koroliana#31' into koroliana#31

# Conflicts:
#	src/main/java/pro/sky/AnimalShelter/service/VolunteerService.java
#	src/main/java/pro/sky/AnimalShelter/utils/MessagesBot.java

koroliana committed:
смерджила в #31 #28-hotfix и #33 Доработка шедулеров

koroliana committed:
реализована команда /all_adopters в меню волонтера
заменен telegram.bot.token на собственный

Microd18 committed:
Создание шедулеров

Microd18 committed:
Merge remote-tracking branch 'origin/develop' into develop

koroliana committed:
конфигурация добавлена в gitignore

koroliana committed:
Пофиксила выдачу уже усыновленного животного и двух животных одному усыновителю.
Добавила @Transactional и javadoc к методам в VolunteerService
Конфигурация добавлена в gitignore

koroliana committed:
Merge pull request #51 from Microd18/#28-hotfix

Добавила проверку OneToOne питомца к усыновителю при усыновлении

Microd18 committed:
#33 Доработка шедулеров

koroliana committed:
Вернула строковые константы для команды /all_adopters в меню волонтера

koroliana committed:
Создала в БД таблицу состояний проверки отчета

DmitriiVolkovIzh committed:
Тесты часть утил и добиты хэндлеры + константы строк

DmitriiVolkovIzh committed:
Merge pull request #52 from Microd18/#29

Тесты часть утил и добиты хэндлеры + константы строк

DmitriiVolkovIzh committed:
CommonUtilsTest 100%

DmitriiVolkovIzh committed:
Merge pull request #53 from Microd18/#29

CommonUtilsTest 100%

Pruglo92 committed:
#34 Покрыть тестами все сервисы которые ещё не покрыты
(покрыт переработан полностью ChatStateServiceTest)

Pruglo92 committed:
Merge branch 'develop' into #34

Pruglo92 committed:
Merge pull request #54 from Microd18/#34

#34 Покрыть тестами все сервисы которые ещё не покрыты

Pruglo92 committed:
#Dev HOTFIX

DmitriiVolkovIzh committed:
Тесты на utils

DmitriiVolkovIzh committed:
Merge pull request #55 from Microd18/#29

Тесты на utils

Pruglo92 committed:
#38 + #40 Сделать возможным запуск меню волонтера /admin без первичной команды /start + Запаролить вход в меню волонтера

Pruglo92 committed:
Merge pull request #56 from Microd18/#38

#38 + #40 Сделать возможным запуск меню волонтера /admin без первично…

koroliana committed:
Реализовала очередь состояний проверки и оценки отчета в меню волонтера

koroliana committed:
Возврат telegram.bot.token

koroliana committed:
Merge branch 'develop' into koroliana#31

koroliana committed:
Merge pull request #31 Проверить и оценить отчет от усыновителя в меню волонтера

При слиянии решены 2 мелких конфликта BotCommand и JsonMapConverter

Pruglo92 committed:
#HotFix фикс для поднятия апликейшена

Pruglo92 committed:
#36 Получить список усыновителей, которые не высылали 2 дня или более отчет WithoutReportAdoptersCommandHandler

Pruglo92 committed:
Merge pull request #58 from Microd18/#36

#36 Получить список усыновителей, которые не высылали 2 дня или более…

Pruglo92 committed:
#36 Получить список усыновителей, которые не высылали 2 дня или более отчет WithoutReportAdoptersCommandHandler FIX

Pruglo92 committed:
Merge pull request #59 from Microd18/#36

#36 Получить список усыновителей, которые не высылали 2 дня или более…

Pruglo92 committed:
#37 Получить список усыновителей, у которых испытательный срок подошел к концу СompletedProbationAdoptersCommandHandler

Pruglo92 committed:
Merge pull request #60 from Microd18/#37

#37 Получить список усыновителей, у которых испытательный срок подоше…

DmitriiVolkovIzh committed:
Тесты на utils

DmitriiVolkovIzh committed:
Merge pull request #61 from Microd18/#29

Тесты на utils
